### PR TITLE
[X86][MC] Add R_X86_64_CODE_6_GOTTPOFF

### PR DESCRIPTION
**Pull request by fzou1**
For

    add %reg1, name@GOTTPOFF(%rip), %reg2
    add name@GOTTPOFF(%rip), %reg1, %reg2
    {nf} add %reg1, name@GOTTPOFF(%rip), %reg2
    {nf} add name@GOTTPOFF(%rip), %reg1, %reg2
    {nf} add name@GOTTPOFF(%rip), %reg

add

  `R_X86_64_CODE_6_GOTTPOFF` = 50

if the instruction starts at 6 bytes before the relocation offset. It's similar to R_X86_64_GOTTPOFF.

Linker can treat `R_X86_64_CODE_6_GOTTPOFF` as `R_X86_64_GOTTPOFF` or convert the instructions above to

    add $name@tpoff, %reg1, %reg2
    add $name@tpoff, %reg1, %reg2
    {nf} add $name@tpoff, %reg1, %reg2
    {nf} add $name@tpoff, %reg1, %reg2
    {nf} add $name@tpoff, %reg

if the first byte of the instruction at the relocation `offset - 6` is `0xd5` (namely, encoded w/REX2 prefix) when possible.

Binutils patch: https://github.com/bminor/binutils-gdb/commit/5bc71c2a6b8efb27089baa1fecded82be4f550a7
Binutils mailthread: https://sourceware.org/pipermail/binutils/2024-February/132351.html
ABI discussion: https://groups.google.com/g/x86-64-abi/c/FhEZjCtDLFw/m/VHDjN4orAgAJ
Blog: https://kanrobert.github.io/rfc/All-about-APX-relocation

### CODE REVIEW

File: llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp [hunk 1]
```diff
@@ -670,6 +672,11 @@ void X86MCCodeEmitter::emitMemModRMByte(
         return Kind == REX2  ? X86::reloc_riprel_4byte_relax_rex2
                : Kind == REX ? X86::reloc_riprel_4byte_relax_rex
                              : X86::reloc_riprel_4byte_relax;
+      case X86::ADD64rm_ND:
+      case X86::ADD64mr_ND:
+      case X86::ADD64mr_NF_ND:
+      case X86::ADD64rm_NF_ND:
+        return X86::reloc_riprel_6byte_relax;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp:679]: You need to consider `ADD64rm_NF` too.

**Comment by KanRobert** [comment]
The description needs to be updated. It doesn't mention NF.

**Comment by fzou1** [comment]
> The description needs to be updated. It doesn't mention NF.

Done. Thanks.

File: llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp (hunk 1 above)
Comment (fzou1) [reply llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp:680]: Added.

**Review by KanRobert — APPROVED** [review]
LGTM

**Comment by fzou1** [comment]
@MaskRay , any comments?

**Review by MaskRay — APPROVED** [review]

**Pull request merged**